fix(server-extensions): rework include documentation

peyerluk opened the pull request:
## Goals

Document what to return in an include renderer when the params are insufficient so that the editor displays no errors and still displays the preview. Also try to fill in missing parts in the docs and try to clarify the use cases by restructuring the docs.

peyerluk commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,71 @@ module.exports = {
+
+  // The ui configuration is used to configure the user interface
+  // in the editor.
+  ui: {
```
@marcbachmann Can you add a description about the configuration possibilites for the `ui` option?

marcbachmann replied:
Here are the types and their configs: https://github.com/upfrontIO/livingdocs-server/blame/master/app/features/includes/api/include_service_factory.js#L19-L43

peyerluk commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,71 @@ module.exports = {
+  },
+  server: {
+    type: 'function',
+    function: function (params, options, callback) {
```
@gabrielhase @mthpvg Do you know what is provided in `options`?

gabrielhase replied:
As far as I remember, the only thing in the options is a parameter indicating if the request comes from the editor or the public API.

marcbachmann replied:
I thought it also contains some kind of params defined in a service in a design.

mthpvg replied:
I confirm what Gabriel said, if the request comes from the editor, `options` is:
```coffeescript
{preview: true}
```
See: https://github.com/upfrontIO/livingdocs-server/blob/master/app/features/includes/controllers/includes_controller.js#L9
If `options` is empty it means the request comes from the public API.

@marcbachmann the `options` object you are mentioning is to my knowledge never sent to the server. It is only used by the editor to toggle UI components:
https://github.com/upfrontIO/livingdocs-design-bluewin/blob/master/source/components/Pages/news-ticker.html#L11

peyerluk commented:
@skloiber-sz Bin jetzt dazu gekommen die Doku aufzuarbeiten, nach unserer Session letze Woche. Die Beschreibung der Render Funktion und der return values ist ganz unten im Dokument. Ich habe gleich noch ein paar andere Dinge überarbeitet.

marcbachmann commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,71 @@ module.exports = {
+    }
+  },
+  server: {
+    type: 'function',
```
There's also a type named `remote`
https://github.com/upfrontIO/livingdocs-server/blame/master/app/features/includes/api/include_service_factory.js#L95-L118

```
server: {
  type: 'remote',
  url: 'https://example.com/render?someDefaultParam=foo'
}
```

peyerluk commented:
@marcbachmann I added include ui documentation and also added an expample with the `options` object in the include render function. Could you quickly review?

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -2,67 +2,76 @@
 
 Think of Livingdocs includes as edge-side includes. You can create include components in your design, using a include [directive](../../livingdocs-framework/directives.md).
```
an include

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
````diff
@@ -2,67 +2,76 @@
-```html
-<ld-include data-include-service="embed-teaser" data-include-params="{"mediaId":2}"></ld-include>
-```
+## A component with an include directive
````
This title does not render in the markdown preview.

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -2,67 +2,76 @@
+Includes are directives on components. They can be used as content placeholders
+that will be filled with HTML created by an include service on the server.
+
+An example component with an include diretive:
```
directive

mthpvg replied:
Would it be useful to add information about the origin of that snippet:
*An example component, defined in the design, with an include directive*

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,134 @@ module.exports = {
+- in the rendered HTML if you choose to resolve the include while rendering
+
+So in order to see a preview while editing you have to register an include renderer
+even if you choose not to resolve includes while rendering.
```
Naive question, is the following statement a real use case:
- I want to see the preview in the editor but I don't want to render it in the public API?

peyerluk replied:
If the include is unresolved it would output this during rendering:

```html
<div>
  <ld-include data-include-service="embed-teaser" data-include-params="{"mediaId":2}"></ld-include>
</div>
```

But in the editor the user should still see the resolved include.

peyerluk replied:
I don't have an idea for a better wording of the sentence in the docs. At least not right now.

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
````diff
@@ -78,4 +87,134 @@ module.exports = {
+}
+```
+
+#### Include UI options:
````
Title is not rendering in the markdown preview.

peyerluk replied:
This happens all the time... :(

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,134 @@ module.exports = {
+        // the include preview.
+        return callback(null, '')
+      } else if (paramsAreInsufficient(params)) {
+        // Return undefined in not enough params are provided to
```
Return undefined because not enough...

mthpvg commented on reference-docs/server-extensions/add-custom-include.md:
```diff
@@ -78,4 +87,134 @@ module.exports = {
+  },
+  server: {
+    type: 'function',
+    function: function (params, options, callback) {
```
What about:
```js
if (paramsAreInsufficient(params) || unHappyCase()) {
  if (options && options.preview === true) return callback(null, undefined)
  return callback(null, '')
}
```
But I could be missing the point.

marcbachmann replied:
Yes, a self-contained example that would work out of the box would be nice.

marcbachmann replied:
Don't call `paramsAreInsufficent`. Just do a manual assertion.
This code can be confusing because of the many unknown variables.

peyerluk replied:
I rewrote the example. I think it makes more sense now.

marcbachmann commented on reference-docs/server-extensions/add-custom-include.md:
````diff
@@ -2,67 +2,76 @@
 ```html
-<div doc-include="embed-teaser"></div>
+<div>
+  <ld-include data-include-service="embed-teaser" data-include-params="{"mediaId":2}"></ld-include>
````
That looks wrong
`data-include-params="{"mediaId":2}"`

peyerluk replied:
Thanks. Fixed. It's `data-include-params="{&quot;mediaId&quot;:2}"`

peyerluk commented:
@mthpvg @marcbachmann Thanks for the feedback ❤️

livingdocs-automation commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/livingdocsIO/documentation/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
